docs/sampler.md

```diff
@@ -0,0 +1,142 @@
+First, download the original trace files from [here](https://azurecloudpublicdataset2.blob.core.windows.net/azurepublicdatasetv2/azurefunctions_dataset2019/azurefunctions-dataset2019.tar.xz) and extract the CSV files (default location: `data/azure/`).
+
+```console
+wget https://azurecloudpublicdataset2.blob.core.windows.net/azurepublicdatasetv2/azurefunctions_dataset2019/azurefunctions-dataset2019.tar.xz -P ./data/azure
+tar -xf ./data/azure/azurefunctions-dataset2019.tar.xz -C data/azure/
+```
+
+Then, run the following command to preprocess the trace.
+
+```console
+python3 -m sampler preprocess -h
+
+usage:  preprocess [-h] [-t path] -o path -s start -dur duration
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -t path, --trace path
+                        Path to the Azure trace files
+  -o path, --output path
+                        Output path for the preprocessed traces
+  -s start, --start start
+                        Time in dd:hh:mm format, at which the excerpt of the postprocessed trace should begin, first day is day 0
+  -dur duration, --duration duration
+                        Duration in minutes of the excerpt extracted from the postprocessed trace             
+```
+
+## Sampling
+
+### Description
+
+The sampling algorithm requires a clean, i.e., preprocessed, trace, which it derives samples.
+The user specifies the size range for samples to be generated, specifying the minimum and maximum sizes
+as well as the step of the sweep. The user also specifies the number of random sampling trials for each sample size.
+For each trial, the sample is evaluated by computing its Wasserstein distance (WD) from the original trace -- for each minute
+in the trace -- in two dimensions, namely the number of invocations in a minute and the amount of CPU and memory resources
+used by the functions. The latter is estimated (at the minute granularity for each function, which then are summed up)
+by multiplying the number of invocations by the average (mean) duration and by the memory footprint.
+
+Note that the latter (resource) WD metric is normalized by 1000 000 to be in the same ballpark as the former (
+invocation) WD metric. The latter metric is found empirically to be 1000 000 higher than the former because function
+invocation duration average is 1000ms, invocation count average is <10/min, and memory average is 200MB.
+
+First, the sampler derives the largest sample, for which it makes several attempts (trials) when choosing the sample
+with the smallest aggregate WD distance (mean of the invocation and resource WDs). Then, each smaller sample is derived
+from the previous, larger sample; hence guaranteeing that large samples always include smaller samples that guarantees
+monotonic load increase (in terms of resource usage) when sweeping the sample size.
+
+## Getting the traces
+
+The reference traces are stored in the `reference_traces` branch of this repo in the root folder, as a `samples.tar.gz`
```

Comment:
This refers to the sampler branch. Will this be added as well or should we just add it to this repo's LFS?